gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/ManifestBasedDataset.java

```diff
@@ -0,0 +1,145 @@
+    return this.manifestPath.toString();
+  }
+
+  @Override
+  public Iterator<FileSet<CopyEntity>> getFileSetIterator(FileSystem targetFs, CopyConfiguration configuration)
+      throws IOException {
+    if (!fs.exists(manifestPath)) {
+      log.warn(String.format("Manifest path %s does not exist on filesystem %s, will not copy data in this manifest"
+          + ", probably due to wrong configuration of %s", manifestPath.toString(), fs.getUri().toString(), ManifestBasedDatasetFinder.MANIFEST_LOCATION));
+      return Iterators.emptyIterator();
+    } else if (fs.getFileStatus(manifestPath).isDirectory()) {
+      log.warn(String.format("Manifest path %s on filesystem %s is a directory, which is not supported. Please set the manifest file locations in"
+          + "%s, you can specify multi locations split by '',", manifestPath.toString(), fs.getUri().toString(), ManifestBasedDatasetFinder.MANIFEST_LOCATION));
+      return Iterators.emptyIterator();
+    }
+    JsonReader reader = null;
+    List<CopyEntity> copyEntities = Lists.newArrayList();
+    List<FileStatus> toDelete = Lists.newArrayList();
+    //todo: put permission preserve logic here?
+    try {
+      reader = new JsonReader(new InputStreamReader(fs.open(manifestPath), "UTF-8"));
+      reader.beginArray();
+      while (reader.hasNext()) {
+        //todo: We can use fileSet to partition the data in case of some softbound issue
+        //todo: After partition, change this to directly return iterator so that we can save time if we meet resources limitation
+        JsonObject file = GSON.fromJson(reader, JsonObject.class);
+        Path fileToCopy = new Path(file.get("fileName").getAsString());
+        if (this.fs.exists(fileToCopy)) {
+          if (!targetFs.exists(fileToCopy) || needCopy(this.fs.getFileStatus(fileToCopy), targetFs.getFileStatus(fileToCopy))) {
+            CopyableFile copyableFile =
+                CopyableFile.fromOriginAndDestination(this.fs, this.fs.getFileStatus(fileToCopy), fileToCopy, configuration)
+                    .fileSet(datasetURN())
+                    .datasetOutputPath(fileToCopy.toString())
+                    .ancestorsOwnerAndPermission(CopyableFile
+                        .resolveReplicatedOwnerAndPermissionsRecursively(this.fs, fileToCopy.getParent(),
+                            new Path("/"), configuration))
+                    .build();
+            copyableFile.setFsDatasets(this.fs, targetFs);
+            copyEntities.add(copyableFile);
+          }
+        } else if (targetFs.exists(fileToCopy)){
+          toDelete.add(targetFs.getFileStatus(fileToCopy));
+        }
+      }
+      if (Boolean.parseBoolean(this.properties.getProperty(DELETE_FILE_NOT_EXIST_ON_SOURCE, "false"))) {
+        CommitStep step = new DeleteFileCommitStep(targetFs, toDelete, this.properties,
+            this.deleteEmptyDirectories ? Optional.of(new Path(this.properties.getProperty(DELETE_EMPTY_DIRECTORIES_UPTO))) : Optional.<Path>absent());
```

Comment:
is this to delete directories that are emptied after we remove the files that do not exist on source?

Reply:
yup, after removing files on dst which are not exist on source, we use this config to determine whether we need to delete the empty dir as well